c++/src/kj/mutex.h

```diff
@@ -607,6 +615,42 @@ inline const T& Lazy<T>::get(Func&& init, const LockSourceLocation& location) co
   return *value;
 }
 
+#if KJ_TRACK_LOCK_BLOCKING
+struct BlockedOnMutexAcquisition {
+  const _::Mutex& mutex;
+  // The mutex we are blocked on.
+
+  const SourceLocation& origin;
+  // Where did the blocking operation originate from.
+};
+
+struct BlockedOnCondVarWait {
+  const _::Mutex& mutex;
+  // The mutex the condition variable is using (may or may not be locked).
+
+  const void* waiter;
+  // Pointer to the waiter that's being waited on.
+
+  const SourceLocation& origin;
+  // Where did the blocking operation originate from.
+};
+
+struct BlockedOnOnceInit {
+  const _::Once& once;
+
+  const SourceLocation& origin;
+  // Where did the blocking operation originate from.
+};
+
+using BlockedOnReason = OneOf<BlockedOnMutexAcquisition, BlockedOnCondVarWait, BlockedOnOnceInit>;
+
+Maybe<BlockedOnReason&> blockedReason() noexcept;
+// Returns the information about the reason the current thread is blocked synchronously on KJ
+// operations. Returns nullptr if the current thread is not currently waiting synchronously on a
+// reason.
```

Comment:
Maybe add: "This is intended to be called from a signal handler."

Otherwise it sounds like it's impossible to return non-null. :)

Reply:
Clarified this section wording further & put a note on why this feature isn't available outside of a signal handler (& how an enhancement could be made that way for anyone who needs it).